invenio_rest/csrf.py

```diff
@@ -0,0 +1,262 @@
+        current_app.config['CSRF_COOKIE_NAME'],
+        _get_new_csrf_token(),
+        max_age=current_app.config.get(
+            'CSRF_COOKIE_MAX_AGE', 60*60*24*7*52),
+        domain=current_app.config.get(
+            'CSRF_COOKIE_DOMAIN',
+            current_app.session_interface.get_cookie_domain(
+                current_app)),
+        path=current_app.session_interface.get_cookie_path(
+                current_app),
+        secure=current_app.config.get('SESSION_COOKIE_SECURE', False),
+        httponly=False,
+        samesite=current_app.config['CSRF_COOKIE_SAMESITE'],
+    )
+
+
+def _get_submitted_csrf_token():
+    header_name = current_app.config['CSRF_HEADER']
+    csrf_token = request.headers.get(header_name)
+    if csrf_token:
+        return csrf_token
+    return None
+
+
+def _is_referrer_secure(referrer):
+    return 'https' in referrer.scheme or \
+        not current_app.config['CSRF_FORCE_SECURE_REFERRER']
+
+
+def _reject(reason):
+        abort(400, reason)
+
+
+def csrf_validate():
+    """Check csrf cookie against request headers."""
+    if request.method not in ('GET', 'HEAD', 'OPTIONS', 'TRACE'):
+        if request.is_secure:
+            referrer = request.referrer
+
+            if referrer is None:
+                return _reject(REASON_NO_REFERER)
+
+            referrer = urlparse(referrer)
+            # Make sure we have a valid URL for Referer.
+            if '' in (referrer.scheme, referrer.netloc):
+                return _reject(REASON_MALFORMED_REFERER)
+
+            # Ensure that our Referer is also secure.
+            if not _is_referrer_secure(referrer):
+                return _reject(REASON_INSECURE_REFERER)
+
+            if referrer.hostname not in current_app.config.get(
+                                            'APP_ALLOWED_HOSTS'):
+                reason = REASON_BAD_REFERER % referrer.geturl()
+                return _reject(reason)
+
+        csrf_token = _get_csrf_token()
+        if csrf_token is None:
+            return _reject(REASON_NO_CSRF_COOKIE)
+
+        request_csrf_token = _get_submitted_csrf_token()
+        if not request_csrf_token:
+            _reject(REASON_BAD_TOKEN)
+
+        decoded_request_csrf_token = _decode_csrf(request_csrf_token)
+
+        if csrf_token != decoded_request_csrf_token:
+            return _reject(REASON_BAD_TOKEN)
+
+
+def reset_token(self):
+        """Change the CSRF token in use for a request."""
+        request.csrf_cookie_needs_reset = True
+
+
+class CSRFTokenMiddleware():
+    """CSRF Token Middleware."""
+
+    def __init__(self, app=None):
+        """Middleware initialization.
+
+        :param app: An instance of :class:`flask.Flask`.
+        """
+        if app:
+            self.init_app(app)
+
+    def init_app(self, app):
+        """Initialize middleware extension.
+
+        :param app: An instance of :class:`flask.Flask`.
+        """
+        app.config.setdefault('CSRF_COOKIE_NAME', '_csrftoken')
```

Comment:
May I ask why the `_`? I understand it is configurable, but it looks like that, e.g. `axios` in javascript, expects `csrftoken` as a name. See here: https://vsupalov.com/avoid-csrf-errors-axios-django/

Reply:
Axios expects to configure the `xsrfCookieName` option. I think that `_` was there just to namespace it I guess. I can prefix it with something like `invenio_csrftoken`. @slint WDYT?

Reply:
Let's go with a default tha works out of the box with common HTTP clients like axios, `csrftoken` it is

Reply:
The defaults in https://github.com/axios/axios are
```
// `xsrfCookieName` is the name of the cookie to use as a value for xsrf token
  xsrfCookieName: 'XSRF-TOKEN', // default

  // `xsrfHeaderName` is the name of the http header that carries the xsrf token value
  xsrfHeaderName: 'X-XSRF-TOKEN', // default
```

Is this fine? Otherwise, leave it as it is :)

Reply:
What I found is: `The difference between the X-CSRF-TOKEN and X-XSRF-TOKEN is that the first uses a plain text value and the latter uses an encrypted value` in **Laravel**!!!
I would go for `_csrftoken` or `csrftoken` or similar...

Reply:
Then I think I will leave it as it is :)

Reply:
Django:
```
CSRF_HEADER_NAME: 'HTTP_X_CSRFTOKEN'
CSRF_COOKIE_NAME: 'csrftoken'
```
https://docs.djangoproject.com/en/3.0/ref/settings/#csrf-header-name